kernel/kernel-api/src/main/java/io/delta/kernel/internal/columndefaults/ColumnDefaults.java

```diff
@@ -0,0 +1,185 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal.columndefaults;
+
+import io.delta.kernel.data.Row;
+import io.delta.kernel.exceptions.KernelException;
+import io.delta.kernel.internal.DeltaErrors;
+import io.delta.kernel.internal.data.TransactionStateRow;
+import io.delta.kernel.internal.util.SchemaIterable;
+import io.delta.kernel.types.*;
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.stream.Stream;
+
+/**
+ * Utilities class for TableFeature "allowColumnDefaults". NOTE: As of Aug 2025, kernel only
+ * supports reading Delta tables with the table feature, or modifying table metadata. Writing actual
+ * data to the table is not allowed.
+ */
+public class ColumnDefaults {
+
+  private static final String DEFAULT_VALUE_METADATA_KEY = "CURRENT_DEFAULT";
+
+  /** Don't allow data writes to tables with default values */
+  public static void blockWriteIfEnabled(Row transactionState) {
+    if (extractFieldsWithDefaultValues(TransactionStateRow.getLogicalSchema(transactionState))
+        .findAny()
+        .isPresent()) {
+      throw new UnsupportedOperationException(
+          "Writing with Column Default values is not supported yet.");
+    }
+  }
+
+  /**
+   * Validate Column Default values in the provided metadata. Kernel only supports literal default
+   * values. See {validateLiteral}.
+   *
+   * @param schema target table schema
+   * @param isEnabled When the feature is disabled, no column default is allowed.
+   * @param isIcebergCompatV3Enabled Kernel currently requires IcebergCompatV3 to be enabled when
+   *     using Column Defaults
+   * @throws KernelException when the table contains invalid default value
+   */
+  public static void validateSchema(
+      StructType schema, boolean isEnabled, boolean isIcebergCompatV3Enabled) {
+    if (isEnabled && !isIcebergCompatV3Enabled) {
+      throw new KernelException(
+          "In Delta Kernel, default values table feature requires IcebergCompatV3 to be enabled.");
+    }
+    Stream<StructField> defaultValues = extractFieldsWithDefaultValues(schema);
+    if (!isEnabled) {
+      if (defaultValues.findAny().isPresent()) {
+        throw new KernelException(
+            "This table does not enable table features for setting column defaults");
```

Comment:
Can you make the error message more clear? Found column defaults in the schema but the table does not support column defaults with the columnDefaults table feature.

Reply:
Also please put in DeltaErrors + a todo to make this an InvalidTableException